[sebin0817] iP

sebin0817 opened the pull request:
# Chatbot Pro 
> "Your mind is for having ideas, not holding them." – David Allen

Chatbot Pro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/admin/ip-w4.html)
2. double-click it
3. add your tasks
4. let it **manage your tasks** for you 😉

`System.out.println("And it is FREE!")`

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Completed Level 0: Rename, Greet, Exit

Commit:
Completed Level 1: Echo

Commit:
Completed Level 2: Add, List

Commit:
Completed Level 3: Mark as Done

Commit:
Completed Level 4: ToDos, Events, Deadlines

Commit:
Completed A-TextUiTesting: Automated Text UI Testing

Commit:
Completed Level-5: Handle Errors

Commit:
Completed Level 6: Delete

Commit:
Completed Level 7: Save

Commit:
Merge branch-Level-7 into master

Commit:
Completed Level 8: Dates and Times

Commit:
Merge branch-Level-8 into master

Commit:
Completed A-MoreOOP: Use more OOP

Commit:
Merge branch-A-MoreOOP into master

Commit:
Completed A-Packages: Organize into Packages

Commit:
Merge branch-A-Packages into master

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Completed A-Gradle: Use Gradle

Commit:
Merge branch A-Gradle into masteR

Commit:
Completed A-JUnit: Add JUnit Tests

Commit:
Merge branch-A-JUnit into master

Commit:
Completed Level 9: Find

Commit:
Merge branch-Level-9 into master

Commit:
Completed Level A-JavaDoc

Commit:
Completed Level A-CodingStandard

Commit:
Merge branch-A-JavaDoc into master

Commit:
Merge branch 'master' of https://github.com/sebin0817/ip into branch-A-CodingStandard

Commit:
Merge branch-A-CodingStandard into master

sebin0817 committed:
Update README.md

wxiaoyun commented:
Overall LGTM, very informative use of names that follow the naming convention. You code is clear and highly readable, keep up the good work.

Commit:
Completed Level A-CheckStyle: User CheckStyle

Commit:
Merge branch-A-CheckStyle into master

Commit:
Merge branch 'master' of https://github.com/sebin0817/ip

marclamp commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+     *
+     * @param keyword User-specified keyword
+     */
+    public void findTasks(String keyword) {
+        ArrayList<Task> foundTasks = new ArrayList<>();
+
+        for (Task task : tasks) {
+            if (task.getInitialDesc().toLowerCase().contains(keyword.toLowerCase())) {
+                foundTasks.add(task);
+            }
+        }
+
+        for (int i = 0; i < foundTasks.size(); i++) {
+            System.out.println((i + 1) + ". " + foundTasks.get(i).getDescription());
+        }
+    }
```
Consider returning an `ArrayList<Task>`, or even better, a `TaskList` object, then using the `Ui` class to print the output. i.e.
`return foundTasks` 
and in the Ui class:
```
public void printList(ArrayList<Task> tasks) {
    for (int i = 0; i < foundTasks.size(); i++) {
            System.out.println((i + 1) + ". " + foundTasks.get(i).getDescription());
}

marclamp commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Lists all tasks stored in tasklist
+     */
+    public void listTask() {
+        for (int i = 0; i < tasksCount; i++) {
+            System.out.println((i + 1) + ". " + tasks.get(i).getDescription());
+        }
+    }
```
Consider shifting this to the `Ui` class to handle the output printing.

marclamp commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,126 @@
+     * For listing out tasks entered by the users
+     * @param tasklist Tasklist with user-specified tasks
+     */
+    public static void listTasks(TaskList tasklist) {
+        out.println(LINE);
+        out.println("Here are the tasks in your list: ");
+        tasklist.listTask();
+        out.println(LINE);
+    }
```
Consider handling the printing of the `Task` objects in this class instead.
```
public static void listTasks(TaskList tasklist) {
    out.println(LINE);
    out.println("Here are the tasks in your list: ");
    for (int i = 0; i < tasksCount; i++) {
        out.println((i + 1) + ". " + tasks.get(i).getDescription());
    }
    out.println(LINE);
}

marclamp commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,183 @@
+                file.createNewFile(); // Create the file if it doesn't exist
+            }
+        } catch (IOException e) {
+            System.out.println("An error occurred while ensuring the task file exists.");
+        }
```
Consider throwing this as a new `ChatbotException` and catching it in your `Chatbot` class
```
public static void ensureFileExists() throws ChatbotException
...
    } catch (IOException e) {
        throw new ChatbotException("An error occurred while ensuring the task file exists.");
    }
}

marclamp reviewed (commented):
LGTM, overall well-organized code. Could do with shifting some of the printing to the Ui class for better organization, and to handle all printing in the Ui class.

Commit:
Completed Level 10: GUI

Commit:
Merge branch-Level-10 into master

Commit:
Complete Level A-Assertions: Using Assertions

Documented important assumptions that should hold at various points in the code.
Any assertion failures detected by runtime can surface bugs and stop the execution of the code.

sebin0817 committed:
Merge pull request #1 from sebin0817/branch-A-Assertions

Complete Level A-Assertions: Using Assertions

Commit:
Improve code quality and refactor code

Examined the code and refactored to improve the code quality where necessary.
This process help increase readability of the code.

sebin0817 committed:
Merge pull request #2 from sebin0817/branch-A-CodeQuality

Improve code quality and refactor code

Commit:
Complete Level C-Sort

Added a new command 'Sort' that sorts tasks stored in the tasklist by keeping Todo tasks at the front, and sorting Event and Deadline by date chronologically thereafter.
This feature helps users to organise their tasks and better manage them.

sebin0817 committed:
Merge pull request #3 from sebin0817/branch-C-Sort

Complete Level C-Sort

Commit:
Complete A-BetterGUI: Better GUI

Improved the GUI to make it more polished.

sebin0817 committed:
Merge pull request #5 from sebin0817/branch-A-BetterGui

Complete A-BetterGUI: Better GUI

Commit:
Update README.md

Commit:
Update UserGuide

sebin0817 committed:
Merge pull request #6 from sebin0817/branch-A-UserGuide

Update UserGuide

sebin0817 committed:
Update README.md

Commit:
Add JUnit tests
